diff -up powertop-1.11/powertop.c.orig powertop-1.11/powertop.c
--- powertop-1.11/powertop.c.orig	2011-08-16 22:25:57.182428428 +0200
+++ powertop-1.11/powertop.c	2011-08-16 22:35:47.943211446 +0200
@@ -37,6 +37,9 @@
 #include <locale.h>
 #include <time.h>
 #include <sys/stat.h>
+#include <errno.h>
+#include <ncurses.h>
+#include <signal.h>
 
 #include "powertop.h"
 
@@ -57,6 +60,8 @@ int topcstate = 0;
 
 int dump = 0;
 
+sigset_t intmask;
+
 #define IRQCOUNT 150
 
 struct irqdata {
@@ -851,6 +856,10 @@ int main(int argc, char **argv)
 	printf("\n\n");
 	print_intel_cstates();
 	stop_timerstats();
+	if (sigemptyset(&intmask) == -1 || sigaddset(&intmask, SIGWINCH) == -1) {
+		perror("Failed to initialize the signal mask");
+		exit(EXIT_FAILURE);
+	}
 
 	while (1) {
 		double maxsleep = 0.0;
@@ -876,9 +885,15 @@ int main(int argc, char **argv)
 
 		key = select(1, &rfds, NULL, NULL, &tv);
 
-		if (key && tv.tv_sec) ticktime = ticktime - tv.tv_sec - tv.tv_usec/1000000.0;
+		if (key > 0 && tv.tv_sec) ticktime = ticktime - tv.tv_sec - tv.tv_usec/1000000.0;
 
+		if (key == -1 && errno != EINTR) {
+			perror("select");
+			exit(EXIT_FAILURE);
+		}
 
+		sigprocmask(SIG_BLOCK, &intmask, NULL);
+		refresh();
 		stop_timerstats();
 		clear_lines();
 		do_proc_irq();
@@ -1044,7 +1059,7 @@ int main(int argc, char **argv)
 		else
 			ticktime = 45;
 
-		if (key) {
+		if (key > 0) {
 			char keychar;
 			int keystroke = fgetc(stdin);
 			if (keystroke == EOF)
@@ -1177,7 +1192,7 @@ int main(int argc, char **argv)
 		memcpy(last_duration, cur_duration, sizeof(last_duration));
 
 
-		
+		sigprocmask(SIG_UNBLOCK, &intmask, NULL);
 	}
 
 	return 0;
